### PR TITLE
list refactor.

### PR DESCRIPTION
**Pull request by xnning**
make List a data type and predefined. Clean up old parts.

```
data List [A] = Nil | Cons A (List [A]);
```

### CODE REVIEW

File: lib/ClosureF.hs [hunk 1]
```diff
@@ -67,12 +67,11 @@ data Expr t e =
    | JNew ClassName [Expr t e]
    | JMethod (Either ClassName (Expr t e)) MethodName [Expr t e] ClassName
    | JField  (Either ClassName (Expr t e)) FieldName ClassName
-   | PolyList [Expr t e] (Type t)
-   | JProxyCall (Expr t e) (Type t)
    | SeqExprs [Expr t e]
    | Data S.RecFlag [DataBind t] (Expr t e)
    | Constr (Constructor t) [Expr t e]
    | Case (Expr t e) [Alt t e]
+   | Error (Type t) (Expr t e)
```
Comment (tomtau) [inline lib/ClosureF.hs:74]: what is this for? why not use JMethod + SeqExprs?

Comment (xnning) [inline lib/ClosureF.hs:74]: this is for throwing error in program. The type is for indicating the return type. The expr is a string for error message. like

```
head (x:List[T]) : T = 
  case x of
     Cons y _ -> y
  |  Nil -> error [T] "head of empty"
```

Comment (tomtau) [inline lib/ClosureF.hs:74]: why can't it be implemented just in the source language and then desugared to JMethod + SeqExprs in Closure F?

Comment (xnning) [inline lib/ClosureF.hs:74]: Sounds great. I will change it later.

Comment (tomtau) [inline lib/ClosureF.hs:74]: I noticed @zhiyuanshi was working on some sort of "Prelude" (predef/IO.sf), so I imagine this is where the error printing could go to

Comment (xnning) [inline lib/ClosureF.hs:74]: @tomtau  Error can return any type without actually making a value of that type.
here is an issue for defining it in f2j : https://github.com/hkuplg/fcore/issues/108

Comment (tomtau) [inline lib/ClosureF.hs:74]: I see. If you have an idea how to fix #108 without adding the extra case in ClosureF, it'd be good. There's going to be the compiler meeting tomorrow (or today already), so we can discuss this issue there.

**Comment by tomtau** [comment]
Looks alright with tests

**Pull request merged**